### PR TITLE
ddl: add schema placement rules

### PR DESCRIPTION
**Pull request by mjonss**
### What problem does this PR solve?

Issue Number: Fixes https://github.com/pingcap/tidb/issues/27972

Problem Summary:

### What is changed and how it works?

What's Changed:
Using the added fields in the schema model (from https://github.com/pingcap/parser/pull/1340 )

How it Works:

### Check List 

Tests

- [X] Unit test
- [X] Integration test

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
Support for placement rules on Schema (default for table/partitions), either named PLACEMENT POLICY or Direct Placement.```

### CODE REVIEW

**Review by xhebox — COMMENTED** [review]
Kind of overlap with #28025, @zhaoxugang :(

**Comment by mjonss** [comment]
@zhaoxugang and I agreed on keeping this PR for the Schema Placement (Create and use as default) and then rebase his PR, https://github.com/pingcap/tidb/pull/28025 , on this and complete for Partitions (and possibly ALTER SCHEMA).

File: executor/show.go [hunk 1]
```diff
@@ -1326,10 +1324,17 @@ func ConstructResultOfShowCreateDatabase(ctx sessionctx.Context, dbInfo *model.D
...
 	}
 	// MySQL 5.7 always show the charset info but TiDB may ignore it, which makes a slight difference. We keep this
 	// behavior unchanged because it is trivial enough.
+	if dbInfo.DirectPlacementOpts != nil {
+		fmt.Fprintf(buf, " %s", dbInfo.DirectPlacementOpts)
+	}
+	if dbInfo.PlacementPolicyRef != nil {
+		fmt.Fprintf(buf, " PLACEMENT POLICY = '%s'", dbInfo.PlacementPolicyRef.Name.O)
+		// FIXME: Support properly quoted policy names in the parser!
+		//fmt.Fprintf(buf, " PLACEMENT POLICY = %s", stringutil.Escape(dbInfo.PlacementPolicyRef.Name.O, sqlMode))
```
Comment (morgo) [inline executor/show.go:1336]: I believe this is fixed in https://github.com/pingcap/parser/pull/1334 ?

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -165,6 +186,7 @@ func (d *ddl) AlterSchema(ctx sessionctx.Context, stmt *ast.AlterDatabaseStmt) (
 		}
 	}
 
+	// TODO: Check if Placement Rules was set/changes? How to handle removal?
```
Comment (morgo) [inline ddl/ddl_api.go:189]: For removal, you don't have to handle. You can't DROP a placement policy if it is in use (simplification we made earlier this week).

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -157,6 +175,9 @@ func (d *ddl) AlterSchema(ctx sessionctx.Context, stmt *ast.AlterDatabaseStmt) (
 				return ErrConflictingDeclarations.GenWithStackByArgs(toCharset, info.CharsetName)
 			}
 			toCollate = info.Name
+			// TODO: Add default and give error?
+			// TODO: Check for all PlacementOpts
```
Comment (morgo) [inline ddl/ddl_api.go:179]: Please cleanup / address this comment before merging. Rest LGTM.

File: executor/show.go (hunk 1 above)
Comment (mjonss) [reply executor/show.go:1336]: Yes, I removed it. It will be fixed in tidb by https://github.com/pingcap/tidb/pull/28055

**Review by xhebox — APPROVED** [review]

**Comment by mjonss** [comment]
@morgo Please take a look

**Review by morgo — APPROVED** [review]

File: ddl/placement_sql_test.go [hunk 4]
```diff
@@ -717,3 +717,88 @@ alter placement policy
...
+	tk.MustExec(`CREATE PLACEMENT POLICY PolicyTableTest LEADER_CONSTRAINTS = "[+region=tl]" FOLLOWER_CONSTRAINTS="[+region=tf]" FOLLOWERS=2 LEARNER_CONSTRAINTS="[+region=tle]" LEARNERS=1`)
+	tk.MustQuery("SHOW PLACEMENT like 'POLICY %PolicySchemaTest%'").Check(testkit.Rows("POLICY PolicySchemaTest LEADER_CONSTRAINTS=\"[+region=nl]\" FOLLOWERS=4 FOLLOWER_CONSTRAINTS=\"[+region=se]\" LEARNERS=4 LEARNER_CONSTRAINTS=\"[+region=be]\" SCHEDULED"))
+	tk.MustQuery("SHOW PLACEMENT like 'POLICY %PolicyTableTest%'").Check(testkit.Rows("POLICY PolicyTableTest LEADER_CONSTRAINTS=\"[+region=tl]\" FOLLOWERS=2 FOLLOWER_CONSTRAINTS=\"[+region=tf]\" LEARNERS=1 LEARNER_CONSTRAINTS=\"[+region=tle]\" SCHEDULED"))
+	tk.MustExec("CREATE SCHEMA SchemaPolicyPlacementTest PLACEMENT POLICY = `PolicySchemaTest`")
+	tk.MustQuery("SHOW CREATE SCHEMA SCHEMAPOLICYPLACEMENTTEST").Check(testkit.Rows("SchemaPolicyPlacementTest CREATE DATABASE `SchemaPolicyPlacementTest` /*!40100 DEFAULT CHARACTER SET utf8mb4 */ PLACEMENT POLICY = `PolicySchemaTest`"))
```
Comment (sylzd) [inline ddl/placement_sql_test.go:742]: @mjonss @morgo Should we keep create schema be compatible with mysql by enclosing /*T[placement]? like :
```sql
CREATE DATABASE `SchemaDirectPlacementTest` /*!40100 DEFAULT CHARACTER SET utf8mb4 */ /*T![placement] PRIMARY_REGION="nl" REGIONS="se,nz" FOLLOWERS=3 */
```

File: ddl/placement_sql_test.go [hunk 5]
```diff
@@ -717,3 +717,88 @@ alter placement policy
...
+
+	tk.MustExec(`CREATE TABLE SchemaDirectPlacementTest.UseSchemaDefault (a int unsigned primary key, b varchar(255))`)
```
Comment (sylzd) [inline ddl/placement_sql_test.go:744]: When I exec the stmt on client, I get error:
```
ERROR 1105 (HY000): failed to notify PD the placement rules: "[PD:placement:ErrRuleContent]invalid rule content, rule '0' from rule group 'TiDB_DDL_291' can not match any store"
```

**Pull request merged**